File: python/src/variant_provider/fixtures.py
"""Shared provider fixtures; no vendor behavior in the runner."""

from pathlib import Path

from jsonschema import Draft202012Validator
from referencing import Registry

from .definition import bundled_bytes, load_definition, strict_json
from .engine import Provider, ReplayTransport


def read_case(path):
    case = strict_json(Path(path).read_bytes())
    schema = strict_json(bundled_bytes("schemas/case-1.0.0.schema.json"))
    Draft202012Validator(schema, registry=Registry()).validate(case)
    return case


def run_case(path, check=True):
    path = Path(path)
    case = read_case(path)
    root = path.parent.parent.parent
    definition, _ = load_definition(root / case["definition"], allow_unsigned=True)
    transport = ReplayTransport(case["calls"])
    environment = dict(case["environment"])
    provider = Provider(definition, transport=transport, environment=environment)
    results = []
    for step in case["steps"]:
        if "environment" in step:
            environment.clear()
            environment.update(step["environment"])
        if step.get("clear_cache", False):
            provider.clear_cache()
        result = provider.invoke(step["method"])
        if check:
            assert result == step["expected"], (
                f"{case['id']} / {step['method']}: {result} != {step['expected']}"
            )
        results.append(result)
    transport.assert_consumed()
    return results


def inventory(directory):
    paths = sorted(Path(directory).glob("*.json"))
    if not paths:
        raise ValueError("No shared cases discovered")
    for path in paths:
        read_case(path)
    return paths

File: rust/src/fixtures.rs
//! Language-neutral conformance case reader, shared by tests and CLI.
use crate::{definition, Provider, ReplayTransport};
use serde_json::{json, Value};
use std::{
    fs,
    path::{Path, PathBuf},
    sync::OnceLock,
};
pub fn read_case(path: &Path) -> Result<Value, String> {
    let case = definition::strict_json(&fs::read(path).map_err(|e| e.to_string())?)?;
    static VALIDATOR: OnceLock<jsonschema::Validator> = OnceLock::new();
    let validator = VALIDATOR.get_or_init(|| {
        jsonschema::draft202012::new(
            &serde_json::from_str(include_str!(concat!(env!("OUT_DIR"), "/case-schema.json")))
                .unwrap(),
        )
        .unwrap()
    });
    validator.validate(&case).map_err(|e| e.to_string())?;
    Ok(case)
}
pub fn inventory(directory: &Path) -> Result<Vec<PathBuf>, String> {
    let mut paths = vec![];
    for entry in fs::read_dir(directory).map_err(|e| e.to_string())? {
        let path = entry.map_err(|e| e.to_string())?.path();
        if path.extension().is_some_and(|e| e == "json") {
            read_case(&path)?;
            paths.push(path);
        }
    }
    paths.sort();
    if paths.is_empty() {
        return Err("No shared cases discovered".into());
    }
    Ok(paths)
}
pub fn run_case(path: &Path, check: bool) -> Result<Vec<Value>, String> {
    let case = read_case(path)?;
    let root = path
        .parent()
        .and_then(Path::parent)
        .and_then(Path::parent)
        .ok_or("Expected json/tests/cases layout")?;
    let d = definition::load(
        &root.join(case["definition"].as_str().unwrap()),
        None,
        &json!([]),
        true,
    )?;
    let mut p = Provider::new(
        d,
        Box::new(ReplayTransport::new(
            case["calls"].as_array().unwrap().clone(),
        )),
    )?;
    p.environment =
        Some(serde_json::from_value(case["environment"].clone()).map_err(|e| e.to_string())?);
    let mut results = vec![];
    for step in case["steps"].as_array().unwrap() {
        if let Some(e) = step.get("environment") {
            p.environment = Some(serde_json::from_value(e.clone()).map_err(|e| e.to_string())?);
        }
        if step["clear_cache"] == true {
            p.clear_cache();
        }
        let result = p.invoke(step["method"].as_str().unwrap());
        if check && result != step["expected"] {
            return Err(format!(
                "{} / {}\nactual: {}\nexpected: {}",
                case["id"], step["method"], result, step["expected"]
            ));
        }
        results.push(result);
    }
    p.finish().map_err(|e| e.to_string())?;
    Ok(results)
}
